python/ray/train/examples/pipeline_parallelism_example.py

```diff
@@ -0,0 +1,352 @@
+        if is_train:
+            data_per_rank = data.size(0) // world_size
+            data = data[rank * data_per_rank : (rank + 1) * data_per_rank]
+        return data.to(device)
+
+    batch_size = 20
+    eval_batch_size = 10
+    train_data = batchify(train_data, batch_size, world_rank, world_size, True)
+    val_data = batchify(val_data, eval_batch_size, world_rank, world_size)
+    test_data = batchify(test_data, eval_batch_size, world_rank, world_size)
+
+    bptt = 35
+
+    def get_batch(source, i):
+        seq_len = min(bptt, len(source) - 1 - i)
+        data = source[i : i + seq_len]
+        target = source[i + 1 : i + 1 + seq_len].view(-1)
+        # Need batch dimension first for pipeline parallelism.
+        return data.t(), target
+
+    ntokens = len(vocab)  # the size of vocabulary
+    emsize = 4096  # embedding dimension
+    nhid = (
+        4096  # the dimension of the feedforward network model in nn.TransformerEncoder
+    )
+    nlayers = 8  # the number of nn.TransformerEncoderLayer in nn.TransformerEncoder
+    nhead = 16  # the number of heads in the multiheadattention models
+    dropout = 0.2  # the dropout value
+
+    from torch.distributed import rpc
+
+    tmpfile = tempfile.NamedTemporaryFile()
+    rpc.init_rpc(
+        name="worker",
+        rank=0,
+        world_size=1,
+        rpc_backend_options=rpc.TensorPipeRpcBackendOptions(
+            init_method="file://{}".format(tmpfile.name),
+            # Specifying _transports and _channels is a workaround and we no longer
+            # will have to specify _transports and _channels for PyTorch
+            # versions >= 1.8.1
+            _transports=["ibv", "uv"],
+            _channels=["cuda_ipc", "cuda_basic"],
+        ),
+    )
+
+    partition_len = ((nlayers - 1) // num_gpus) + 1
+
+    # Add encoder in the beginning.
+    tmp_list = [Encoder(ntokens, emsize, dropout).cuda(num_gpus * local_rank)]
+    module_list = []
+
+    # Add all the necessary transformer blocks.
+    for i in range(nlayers):
+        transformer_block = TransformerEncoderLayer(emsize, nhead, nhid, dropout)
+        if i != 0 and i % (partition_len) == 0:
+            module_list.append(nn.Sequential(*tmp_list))
+            tmp_list = []
+        device = i // (partition_len)
+        tmp_list.append(transformer_block.to(num_gpus * local_rank + device))
+
+    # Add decoder in the end.
+    tmp_list.append(Decoder(ntokens, emsize).cuda(num_gpus * local_rank + num_gpus - 1))
+    module_list.append(nn.Sequential(*tmp_list))
+
+    # Need to use 'checkpoint=never' since as of PyTorch 1.8, Pipe checkpointing
+    # doesn't work with DDP.
+    from torch.distributed.pipeline.sync import Pipe
+
+    chunks = 8
+    model = Pipe(torch.nn.Sequential(*module_list), chunks=chunks, checkpoint="never")
+
+    # Initialize process group and wrap model in DDP.
+    from torch.nn.parallel import DistributedDataParallel
+
+    model = DistributedDataParallel(model)
```

Comment:
use torch.prepare_model() here?

Reply:
Yeah, let me add that with `move_to_device=False`.

Reply:
Actually this doesn't work:
```
ValueError: DistributedDataParallel device_ids and output_device arguments only work with single-device/multiple-device GPU modules or CPU modules, but got device_ids [0], output_device 0, and module parameters {device(type='cuda', index=0), device(type='cuda', index=1)}.
```